### PR TITLE
Addon Manager: refactor process_string_to_datetime

### PR DESCRIPTION
**Pull request by chennes**
Move it to the utilities file and get it under test.

### CODE REVIEW

**Comment by chennes** [comment]
> The title on this commit "Refactor utilities tests to remove filesystem use" seems unrelated to the content?

It looks like I mis-squashed the various commits here: I'll pull it back apart and do it again -- thanks for the catch.

**Comment by chennes** [comment]
> And finally, as an old dude, I can confirm that years before 2001 are real.

But it's not possible for a FreeCAD macro to have been written before then :).

**Comment by chennes** [comment]
Massaging that code into non-pseudocode that passes the tests gives something like:
```
def process_date_string_to_python_datetime(string: str) -> datetime:

    could_be_year = lambda x: x.isnumeric() and int(x) > 2000  # Addon creation year must be at least 2001, or it's clearly a mistake
    could_be_month = lambda x: 1 <= int(x) <= 12

    vals = re.split(r"[ ./-]+", string.strip())  # Leaves the original regex alone so we don't break existing macros

    if len(vals) == 3:
        try:
            pos = [] # positions of d, m, y
            if could_be_year(vals[0]):
                pos = [0,1,2]
            elif could_be_year(vals[2]):
                pos = [2,1,0] if could_be_month(vals[1]) else [2,0,1]
            vals = [int(x) for x in vals] # converted to ints
            return datetime(vals[pos[0]], vals[pos[1]], vals[pos[2]])
        except (IndexError, ValueError):
            pass
    raise ValueError(f"'{string}' could not be interpreted as a date")
```
I'll say that when doing this fix, my general feeling is that using the double-index method made it more difficult to reason about, and it feels a bit C-ish. I had to step through it with the debugger to see the error with the indexing on the line that now reads `pos = [2,1,0] if could_be_month(vals[1]) else [2,0,1]`. I'm also not overly excited about deliberately allowing an `IndexError` and catching it, but that's a sort of stylistic argument. I'll leave it to the merging Maintainer to decide if this proposed change is worth making. For reference, the current code is:
```
def process_date_string_to_python_datetime(date_string: str) -> datetime:

    def raise_error(bad_string: str):
        raise ValueError(f"Unrecognized date string '{bad_string}' (expected YYYY-MM-DD or DD-MM-YYYY)")

    split_result = re.split(r"[ ./-]+", date_string.strip())
    if len(split_result) != 3:
        raise_error(date_string)

    try:
        split_result = [int(x) for x in split_result]
        # The earliest possible year an addon can be created or edited is 2001:
        if split_result[0] > 2000:
            return datetime(split_result[0], split_result[1], split_result[2])
        elif split_result[2] > 2000:
            # Generally speaking it's not possible to distinguish between DD-MM and MM-DD, so try the first, and
            # only if that fails try the second
            if split_result[1] <= 12:
                return datetime(split_result[2], split_result[1], split_result[0])
            return datetime(split_result[2], split_result[0], split_result[1])
        else:
            raise ValueError(f"Invalid year in date string '{date_string}'")
    except ValueError:
        raise_error(date_string)
```

**Comment by maxwxyz** [comment]
@jbaehr wants to take a look at this

**Comment by jbaehr** [comment]
I personally prefer the current proposal of this PR. Yes, we have *three times* the datetime construction interleaved with date-format-guessing, but this all happens within only *eight lines of code* (including comments). On that scale, reasoning about the *one thing* the methods does (fuzzy date parsing) is quite straight forward and does not require further separation. With the int-parsing out of the way, I like it more than the double-indexing, which by itself adds more cognitive load (for me, the tipping point where the solution complexity outweighs the problem complexity is reached here).

Engineering is always about tradeoffs, and software engineering is no exception here. On a larger scale, I consider separation of concerns as fundamentally important. Within a single short method, I value local readability higher.

---

I do have another remark, though: When handling the ValueError, you raise another ValueError. This one, however, does not happen *while handling* the original error. Rather it's a *direct consequence* of the original error. This is something the code should express. (Python calls this [Exception Chaining](https://docs.python.org/3/tutorial/errors.html#exception-chaining))

```diff
-    def raise_error(bad_string: str):
-        raise ValueError(f"Unrecognized date string '{bad_string}' (expected YYYY-MM-DD)")
+    def raise_error(bad_string: str, root_cause: Exception = None):
+        raise ValueError(f"Unrecognized date string '{bad_string}' (expected YYYY-MM-DD)") from root_cause
@@
-    except ValueError:
-        raise_error(date_string)
+    except ValueError as exc:
+        raise_error(date_string, exc)
```

**Comment by maxwxyz** [comment]
and there are also two conflicts

**Comment by chennes** [comment]
Fortunately I don't have to support an open set of any possible date formats: we now require the use of ISO 8601 formatting in new macros, so only the set of existing macros needs to be supported. The proposed code covers all of those cases. Eventually it would be nice to go back through the old macros and update them, but I haven't had the time or inclination to deal with that yet.

File: src/Mod/AddonManager/AddonManagerTest/app/test_utilities.py [hunk 1]
```diff
@@ -132,21 +131,22 @@ def test_get_assigned_string_literal(self):
             result = get_assigned_string_literal(line)
             self.assertIsNone(result)
 
-    def test_get_macro_version_from_file(self):
-        if FreeCAD:
-            test_dir = os.path.join(
-                FreeCAD.getHomePath(), "Mod", "AddonManager", "AddonManagerTest", "data"
-            )
-            good_file = os.path.join(test_dir, "good_macro_metadata.FCStd")
```
Comment (jbaehr) [inline src/Mod/AddonManager/AddonManagerTest/app/test_utilities.py:140]: If this test file is not needed any longer, it should be removed from the source tree (dito for bad_macro_metadata.FCStd below)

Comment (chennes) [reply src/Mod/AddonManager/AddonManagerTest/app/test_utilities.py:140]: Thanks for the reminder, I've removed them now.

**Comment by chennes** [comment]
(Note to maintainers: I only left the commit history intact to make the review process clearer, this can get squashed on merge).

**Comment by jbaehr** [comment]
LGTM

**Pull request merged**